### PR TITLE
broadcast rewrite with custom functinos

### PR DESCRIPTION
**Pull request by chengchingwen**
As mentioned in #253, support `@cufunc` to be used with some custom functions that had already registered.

### CODE REVIEW

File: test/base.jl [hunk 1]
```diff
@@ -77,6 +77,24 @@ end
   @test Array(Whatever{Int}.(CuArray([1]))) == Whatever{Int}.([1])
 end
 
+@testset "Cufunc" begin
+  import CUDAnative
```
Comment (maleadt) [inline test/base.jl]: why is this needed?

**Comment by maleadt** [comment]
bors try

File: test/base.jl (hunk 1 above)
Comment (chengchingwen) [reply test/base.jl]: you are right, I forget that I change the implementation.

File: src/broadcast.jl [hunk 2]
```diff
@@ -39,9 +39,12 @@ end
 
 using MacroTools
 
+const _cufuncs = copy(libdevice)
+cufuncs() = (global _cufuncs; _cufuncs)
+
 function replace_device(ex)
   MacroTools.postwalk(ex) do x
-    x in libdevice ? :(CUDAnative.$x) : x
+    x in cufuncs() ? :(CuArrays.cufunc($x)) : x
```
Comment (MikeInnes) [inline src/broadcast.jl]: You could replace this with `x in _cufuncs` and get rid of the indirection here
